### ISSUE TITLE
[bug] Forking an effect creator causes it to run

### ISSUE THREAD

**Issue by Andarist** [redux-saga/redux-saga#1130]
Reproduced issue: https://codesandbox.io/s/68rm79Qg7

**yield fork(call, fn)**
its ok, as `call` is just a function so it shouldnt be treated any different

**yield fork(() => call(fn))**
its not ok, `fn` gets executed, returned effect is interpreted, it should not happen, may lead to unexpected behaviours

**yield fork(takeEvery, 'pattern', fn)**
its not ok, `takeEvery` uses `fork` under the hood, and it seems that nested `fork(fork, fn, ...)` gets 'unwrapped' and unwrapped task descriptor (one for `takeEvery('pattern', fn)` / `fork(fn, ...)`) gets returned to the caller saga.

Probably similar thing happens when returning iterators - need to be checked.

Returning promises / iterators from `call` is working, but should be treated as special cases.

**Comment by Andarist** [comment]
**yield fork(() => call(fn))** case need to be checked further. But I suspect its as simple as - returned value gets passed to the interpreter.

Im still a little bit unsure if it should get interpreted. On one hand its just a "yielded" value with a `{done: true}`, OTOH its a little bit magical and might be unexpected, its quite implicit behaviour.

I.e. for of loops ignore the "done" values, but i.e. Promises auto-flatten themselves when chained (also not sure if that behaviour is good on Promises)

**Comment by Andarist** [comment]
somewhat related https://github.com/TheNavigateur/proposal-pipeline-operator-for-function-composition/issues/3

**Comment by nhducit** [comment]
currently we can use **yield fork(call, fn)** instead of **yield fork(() => call(fn))** to avoid this bug. How about **yield fork(takeEvery, 'pattern', fn)**? Do we have any workaround solution?

**Comment by Andarist** [comment]
Just don't use any of those, those effects are not intended to be used like that.

**Comment by restrry** [comment]
I think problem with `call` is related to [this problem as well](https://github.com/redux-saga/redux-saga/pull/1017). we need to align approach for both cases @shinima @Andarist

**Comment by shinima** [comment]
> Just don't use any of those, those effects are not intended to be used like that.  👍

We should always **create a forking effect** instead of **forking an effect creator**.


Why `fn` in `yield fork(() => call(fn))` get executed is because that we are using [`createTaskIterator`](https://github.com/redux-saga/redux-saga/blob/f85674b3a319c1f9c209e62bc509a0e2c0b4bdeb/packages/core/src/internal/proc.js#L513) to create the task-iterator for fork effect. When [`createTaskIterator` meets a function that returns non-iterator value](https://github.com/redux-saga/redux-saga/blob/f85674b3a319c1f9c209e62bc509a0e2c0b4bdeb/packages/core/src/internal/proc.js#L134-L148), it uses `makeIterator` to create an iterator object as follows:

```js
makeIterator(
  (function() {
    let pc
    const eff = { done: false, value: result }
    const ret = value => ({ done: true, value })
    return arg => {
      if (!pc) {
        pc = true
        return eff
      } else {
        return ret(arg)
      }
    }
  })(),
)
```

This code could be simplified if we use ES2015:
```js
(function* iteratorYieldReturn(value) {
  return yield value // notice the `yield` keyword
})()
```

Notice the `yield` keyword in above code, this means in our example `call(fn)` will be yielded and translated by the saga-middleware. So `fn` gets executed.

The `return yield value` may seem like a trick at first glance. But after a second thought, it is NEEDED so that we could fork an promise (e.g. `fork(fetch, 'some-url')`) or other types of effect (e.g. all/race).

OTOH, `call` effects have extra logic when `fn` returns promise/iterator, so using call and fork effect simultaneously really makes things hard to understand. I think we could mention in our doc that it's a bad idea to use call/fork/takeEevery at the same time.

**Comment by restrry** [comment]
the open question is should we support the same behaviour for returned values? if generator finishes with effect descriptor, like `return call(..)`  we  will receive `{done: true, value: call(...) }`. should we interpret it (as for mentioned `yield call / yield (() => call(..))`) or should they get returned to the caller as is?

**Comment by shinima** [comment]
> if generator finishes with effect descriptor, like return call(..) we will receive {done: true, value: call(...) }. should we interpret it?

We need not interpret returned values from generators. Since if the users want effect descriptors to be interpreted, users could simply add a `yield` as follows:

```diff
function* gen() {
- return call(fetch, 'some-url')
+ return yield call(fetch, 'some-url')
}
```

---

I think the existing behaviors of fork/call are fine. We could add suggestions in docs to prevent the users from using fork and call at the same time.

**Comment by restrry** [comment]
yes, that's easy to fix in user land, but there is undocumented inconsistency in `yield` and `return` behaviour. at least some ppl expect them to handle nested effects in common way https://github.com/redux-saga/redux-saga/issues/1012
as for me - I'm not sure what's desired behaviour for such case and probably keep API without changing, just update docs

**Comment by Andarist** [comment]
I would vote on ignoring "done" values and passing them "as is" to the caller - flattening returned values being not automatic leads to more flexibility. As mentioned adding `yield` is no big deal and if we return such effects to the callers new patterns might emerge - its up to the caller to run (yield) the effect and it might not be instantaneous so one could implement some kind of "merging" strategies, similar to concatMap/mergeMap/exhaustMap/switchMap from stream's world.

**Comment by restrry** [comment]
so the only action point here is add notice in the docs that we shouldn't use `call / fork / take* ` at the same time. correct? that could be done in https://github.com/redux-saga/redux-saga/issues/1478

**Comment by Andarist** [comment]
Actually we should check if something can be done in `createTaskIterator` to prevent  **yield fork(() => call(fn))** from running. Also why `yield fork(takeEvery, 'pattern', fn)` is getting flattened and thus executed too.

Both are rather low priority though.

**Comment by lega0208** [comment]
Sorry if this isn't an appropriate thread to ask, but it's on-topic and I figured it was better than to post another issue.
I'm refactoring a project to use redux-saga and so far it's simplified thing a lot. I appreciate all the work you guys are doing on this. However, I'm having some issues now, and the errors don't bubble up, so I have no way of knowing what's actually causing the errors, but I can see you're already working on errors. 
**To get the the point:**

**Comment by lega0208** [comment]
Sorry if this isn't an appropriate thread to ask, but it's on-topic and I figured it was better than to post another issue.
I'm refactoring a project to use redux-saga and so far it's simplified thing a lot. I appreciate all the work you guys are doing on this. However, I'm having some issues now, and the errors don't bubble up, so I have no way of knowing what's actually causing the errors, but I can see you're already working on errors. 
**To get the the point:**

**Comment by lega0208** [comment]
Sorry if this isn't an appropriate thread to ask, but it's on-topic and I figured it was better than to post another issue.

I'm refactoring a project to use redux-saga and so far it's simplified thing a lot. I appreciate all the work you guys are doing on this. However, I'm having some issues now.

I think my issues are stemming from what you're talking about here. The problem is, I don't know what *not* to do. The solution above of "Just don't use any of those" may seem obvious to you if you're familiar with the codebase, but to someone like me who 1. isn't that experienced, and 2. isn't familiar with the code base, this isn't obvious at all.

@shinima mentions above that we should avoid using call/fork/takeEvery at the same time, but obviously we need to use all of these for different reasons. So in what specific situations should you not use them at the same time? Also how are we supposed to properly nest our effects?
For example this is how I've structured my sagas:

- My rootSaga yields an `all()` effect with an array of all my top-level "watchers", each wrapped in a `fork()`. If I understand correctly (I probably don't), I should fork them all so that they can all watch for actions in parallel.
- My top-level "watchers" watch for actions for each of my pages.
- My page "watchers" yield an `all()` with an array of all my "watchers" for the specific page, again wrapped with `fork()`s.
- Those watchers then implement the `while (true) -> yield take(pattern) -> yield call()/put()` pattern.

Does that make any sense? Or is this a good example of how *not* to structure your sagas?
If this isn't right, how would you suggest structuring them? (Maybe a "Structuring sagas" section could be added to the docs?)

Thanks for your time! Not just for reading/answering my question but for also for building this fantastic tool for all of us. It really is appreciated, you guys are the best.

(Also, sorry if you got multiple notifications from me posting, IE11 doesn't play well with github and so trying to add a line break resulted in me posting prematurely. Gotta love having to use outdated technologies.)

**Comment by Andarist** [comment]
@lega0208 everything you have described sounds OK, you are not using any weird patterns.

This issue is solely about weird code like `yield fork(() => call(someSaga))`, `yield fork(call, someSaga)` and similar. They currently work purely by accident, they don't make much sense and will be "forbidden" in the future in a sense that those "inner" effects wont get executed.

**Comment by Andarist** [comment]
Don't ever be afraid of contributing back to OSS. Most of the project maintainers are really helpful in reviewing and explaining the internals.

**Comment by Andarist** [comment]
Fixed by #1573
